server/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQExclusiveConsumer.java

```diff
@@ -0,0 +1,174 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.rabbitmq.AcknowledgableDelivery;
+import reactor.rabbitmq.ConsumeOptions;
+import reactor.rabbitmq.RabbitFluxException;
+import reactor.rabbitmq.ReceiverOptions;
+
+/**
+ * Taken from {@link reactor.rabbitmq.Receiver}
+ * In order to be able to set the `exclusive` parameter to `true`
+ * to the `channel.basicConsume` method.
+ *
+ * @deprecated to remove once the parallel execution of task has been implemented
+ */
+@Deprecated
+public class RabbitMQExclusiveConsumer implements Closeable {
+    private static final Function<Connection, Channel> CHANNEL_CREATION_FUNCTION = new RabbitMQExclusiveConsumer.ChannelCreationFunction();
+    private static final Logger LOGGER = LoggerFactory.getLogger(RabbitMQExclusiveConsumer.class);
+
+    private static final boolean NON_LOCAL = true;
+    private static final boolean EXCLUSIVE = true;
+
+    private static class ChannelCreationFunction implements Function<Connection, Channel> {
+
+        @Override
+        public Channel apply(Connection connection) {
+            try {
+                return connection.createChannel();
+            } catch (IOException e) {
+                throw new RabbitFluxException("Error while creating channel", e);
+            }
+        }
+    }
+
+    private Mono<? extends Connection> connectionMono;
+    private final AtomicBoolean hasConnection;
+    private final Scheduler connectionSubscriptionScheduler;
+    private final boolean privateConnectionSubscriptionScheduler;
+
+    public RabbitMQExclusiveConsumer(ReceiverOptions options) {
+        this.privateConnectionSubscriptionScheduler = options.getConnectionSubscriptionScheduler() == null;
+        this.connectionSubscriptionScheduler = options.getConnectionSubscriptionScheduler() == null ?
+            createScheduler("rabbitmq-receiver-connection-subscription") : options.getConnectionSubscriptionScheduler();
+        hasConnection = new AtomicBoolean(false);
+        this.connectionMono = options.getConnectionMono() != null ? options.getConnectionMono() :
+            Mono.fromCallable(() -> options.getConnectionFactory().newConnection())
+                .doOnSubscribe(c -> hasConnection.set(true))
+                .subscribeOn(this.connectionSubscriptionScheduler)
+                .cache();
+    }
+
+    protected Scheduler createScheduler(String name) {
+        return Schedulers.newElastic(name);
+    }
+
+
+    public Flux<AcknowledgableDelivery> consumeExclusiveManualAck(final String queue, ConsumeOptions options) {
+        // TODO track flux so it can be disposed when the sender is closed?
+        // could be also developer responsibility
+        return Flux.create(emitter -> connectionMono.map(CHANNEL_CREATION_FUNCTION).subscribe(channel -> {
```

Comment:
I didn't noticed this ugly method during previous review but it doesn't match our usual quality principles: there are too many states and operations for a single method.
Could you try to extract some parts to make the intent more obvious?

Reply:
It's the code from the reactor rabbit `Receiver` class.
The only changes we did, were to set the `exclusive` parameter to true when calling the basic consume method. 
It is deprecated and it's clearly stated in the comments at the class level.

Reply:
are you confident to had some behaviors to such a complex code? Is the test harness good enough?

Reply:
For the life expectancy of this code ( some weeks), yes.